flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java

```diff
@@ -123,46 +126,47 @@ void testExceptionDuringReload() throws Exception {
+    /**
+     * Cache loader creates additional threads in case of multiple input splits. In both cases cache
+     * loader must correctly react on close and interrupt all threads.
+     */
+    @ParameterizedTest
+    @MethodSource("numSplits")
+    void testCloseDuringReload(int numSplits) throws Exception {
+        OneShotLatch reloadLatch = new OneShotLatch();
         Runnable reloadAction =
-                ThrowingRunnable.unchecked(
-                        () -> {
-                            sleepCounter.incrementAndGet();
-                            Thread.sleep(1000);
-                        });
-        InputFormatCacheLoader cacheLoader = createCacheLoader(0, reloadAction);
+                () -> {
+                    reloadLatch.trigger();
+                    assertThatThrownBy(() -> new OneShotLatch().await())
+                            .as("Wait should be interrupted if everything works ok")
+                            .isInstanceOf(InterruptedException.class);
+                    Thread.currentThread().interrupt(); // restore interrupted status
+                };
         InterceptingCacheMetricGroup metricGroup = new InterceptingCacheMetricGroup();
-        cacheLoader.open(metricGroup);
-
-        // check interruption
-        ExecutorService executorService = Executors.newSingleThreadExecutor();
-        Future<?> future = executorService.submit(cacheLoader);
-        executorService.shutdownNow(); // internally interrupts a thread
-        assertThatNoException().isThrownBy(future::get); // wait for the end
-        // check that we didn't process all elements, but reacted on interruption
-        assertThat(sleepCounter).hasValueLessThan(totalSleepCount);
+        CompletableFuture<Void> future;
+        try (InputFormatCacheLoader cacheLoader =
+                createCacheLoader(numSplits, DEFAULT_DELTA_NUM_SPLITS, reloadAction)) {
+            cacheLoader.initializeMetrics(metricGroup);
+            future = cacheLoader.reloadAsync();
+            reloadLatch.await(5000, TimeUnit.MILLISECONDS);
+        }
+        // try-with-resources calls #close, which should wait for the end of reload
+        assertThat(future.isDone()).isTrue();
+        assertThat(metricGroup.loadCounter.getCount()).isEqualTo(1);
         assertThat(metricGroup.numLoadFailuresCounter.getCount()).isEqualTo(0);
+    }
 
-        sleepCounter.set(0);
-
-        // check closing
-        executorService = Executors.newSingleThreadExecutor();
-        future = executorService.submit(cacheLoader);
-        cacheLoader.close();
-        assertThatNoException().isThrownBy(future::get); // wait for the end
-        // check that we didn't process all elements, but reacted on closing
-        assertThat(sleepCounter).hasValueLessThan(totalSleepCount);
-        assertThat(metricGroup.numLoadFailuresCounter.getCount()).isEqualTo(0);
+    static Stream<Arguments> numSplits() {
+        return Stream.of(Arguments.of(1), Arguments.of(2));
```

Comment:
AFAIU, `numSplits` being `1` means that there is no extra fixed-size thread pool created in the `InputFormatCacheLoader`, but for `2` we expect additional threads to be created. Is this considered an implementation detail or shall we test this feature through some `@VisibleForTesting` utility method in `InputFormatCacheLoader` that returns the threadpool (which would be `null` in case of `numSplits=1`). :thinking:

Reply:
I think it can be implementation detail, especially considering that [FLINK-30354](https://issues.apache.org/jira/browse/FLINK-30354) can significantly change internal logic